### PR TITLE
Dev/xygu/20230823/shadow-rework

### LINKED ISSUE #707: [WinUI/ShadowContainer] ThemeResources used inside ShadowContainer are not changing when switching themes
## Current behavior

ThemeResources used inside ShadowContainer are not changing when switching themes in the sample app. If the theme switching is done using `RequestedTheme` instead in App.xaml the issue is not present.

## Expected behavior

ThemeResources used inside ShadowContainer should be properly updated when switching themes

## How to reproduce it (as minimally and precisely as possible)

- Current behavior when switching to Dark Theme with the snippet below 
image

- Expected behavior when switching to Dark Theme with the snippet below 
(can be tested directly using RequestedTheme=Dark in App.xaml)
image

Snippet available on this branch: https://github.com/unoplatform/uno.toolkit.ui/tree/dev/agzi/ShadowContainerThemeSwitchingIssue

```
<!-- Test issue Theme Switching -->
							<Border>
								<Border.Resources>
									<ResourceDictionary>
										<ResourceDictionary.ThemeDictionaries>
											<!-- Shadow Styles ColorPaletteOverride -->
											<ResourceDictionary x:Key="Dark">
												<Color x:Key="OnSurfaceColor">#E6E1E5</Color>
												<Color x:Key="OnSurfaceVariantColor">#C9C5D0</Color>
												<Color x:Key="PrimaryColor">#C7BFFF</Color>

												<SolidColorBrush x:Key="OnSurfaceBrush"
																 Color="{StaticResource OnSurfaceColor}" />

												<SolidColorBrush x:Key="OnSurfaceVariantBrush"
																 Color="{StaticResource OnSurfaceVariantColor}" />

												<SolidColorBrush x:Key="PrimaryBrush"
																 Color="{StaticResource PrimaryColor}" />
												
												<!-- Shadow colors-->
												<Color x:Key="LightShadowColor">#3C3846</Color>
												<Color x:Key="ShadeShadowColor">#24222A</Color>
												
											</ResourceDictionary>
											<ResourceDictionary x:Key="Light">
												<Color x:Key="OnSurfaceColor">#1C1B1F</Color>
												<Color x:Key="OnSurfaceVariantColor">#8A8494</Color>
												<Color x:Key="PrimaryColor">#5946D2</Color>

												<SolidColorBrush x:Key="OnSurfaceBrush"
																 Color="{StaticResource OnSurfaceColor}" />

												<SolidColorBrush x:Key="OnSurfaceVariantBrush"
																 Color="{StaticResource OnSurfaceVariantColor}" />

												<SolidColorBrush x:Key="PrimaryBrush"
																 Color="{StaticResource PrimaryColor}" />

												<!-- Shadow colors-->
												<Color x:Key="LightShadowColor">#FFFFFF</Color>
												<Color x:Key="ShadeShadowColor">#AEAEC0</Color>
											</ResourceDictionary>
										</ResourceDictionary.ThemeDictionaries>

										<utu:ShadowCollection x:Key="NeumorphismBulgingTest">
											<utu:Shadow BlurRadius="10"
														IsInner="True"
														OffsetX="-5"
														OffsetY="-5"
														Opacity="1"
														Spread="0"
														Color="{ThemeResource LightShadowColor}" />
											<utu:Shadow BlurRadius="10"
														IsInner="True"
														OffsetX="4"
														OffsetY="4"
														Opacity="1"
														Spread="0"
														Color="{ThemeResource ShadeShadowColor}" />
										</utu:ShadowCollection>
									</ResourceDictionary>
								</Border.Resources>

								<utu:ShadowContainer Shadows="{StaticResource NeumorphismBulgingTest}">
									<Button Width="80"
											Height="80"
											Background="{ThemeResource OnSurfaceBrush}"
											BorderBrush="{ThemeResource OnSurfaceVariantBrush}"
											Content="Circular"
											CornerRadius="40"
											Foreground="{ThemeResource PrimaryBrush}" />
								</utu:ShadowContainer>
							</Border>
```

Nuget Package: Uno.Toolkit.UI 

Package Version(s): 4.0.0-dev.67

Affected platform(s):

- [X] Windows
- [?] Double-check properly the other platforms

IDE:

- [X] Visual Studio 2022
- [ ] Visual Studio 2019
- [ ] Visual Studio Code
- [ ] Visual Studio for Mac
- [ ] Rider Windows
- [ ] Rider macOS

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): closes #707

## PR Type

What kind of change does this PR introduce?
- Bugfix
- Refactoring (no functional changes, no api changes)

## What is the new behavior?
- content's background not responsive to theme changes
- major refactor

## PR Checklist

Please check if your PR fulfills the following requirements:
- [x] Tested code with current [supported SDKs](../README.md#supported)
- [x] Tested the changes where applicable:
	- [ ] UWP
	- [ ] WinUI
	- [ ] iOS
	- [ ] Android
	- [ ] WASM
	- [ ] MacOS
- [ ] Updated the documentation as needed:
	- [ ] [General Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc)
	- [ ] [Controls Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc/controls)
	- [ ] [Extensions Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc/helpers)
	- [ ] [controls-styles.md](https://github.com/unoplatform/uno.toolkit.ui/blob/main/doc/controls-styles.md)
	- [ ] [lightweight-styling.md (LightWeight Styling Resource Keys)](https://github.com/unoplatform/uno.toolkit.ui/blob/main/doc/lightweight-styling.md)
- [ ] [Runtime Tests and/or UI Tests](https://platform.uno/docs/articles/contributing/guidelines/creating-tests.html) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information
n/a

### CODE REVIEW

**Comment by Xiaoy312** [comment]
saving top comment

File: src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs [hunk 1]
```diff
@@ -228,26 +210,15 @@ private void OnSurfacePainted(object? sender, SKPaintSurfaceEventArgs e)
...
-			}
-
-			TrySetContentBackground(new SolidColorBrush(Color.FromArgb(0, 0, 0, 0)));
-			return true;
-		}
+			SolidColorBrush scb => scb.Color,
 
-		return false;
+			null => default(Color?),
+			_ => throw new NotSupportedException($"Invalid background brush type: {background.GetType().Name}")
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs:220]: it is probably VERY BAD to throw inside a paint procedure of skia
we can probably just treat it as transparent and continue as is
and, maybe just log at err-level here
\---
we are already throwing during the setup with `BindToPaintingProperties`

**Review by roubachof — CHANGES_REQUESTED** [review]
I'm reviewing from a iPhone, and there are a lot of changes here so I'm only addressing an obvious issue: why remove the detection of child corner radius?

For now we only support children with corner radius, but later we could support text or images. In those cases the path for the shadows will be detected just as we detect children corner radius.
We need to determine the aspect of the shadows from the child.

**Comment by Xiaoy312** [comment]
it all got moved to .ShadowContainer.cs as BindToPaintingProperties
and i refactored it to be cleaner and have all monitors (corner radius, background, shadows) in one place

$class-name.properties.cs should stay pure, nothing more than dependency property definitions

File: src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs [hunk 2]
```diff
@@ -122,155 +46,131 @@ private void OnSurfacePainted(object? sender, SKPaintSurfaceEventArgs e)
...
-			// If a property has changed dynamically, we don't want to cache the updated shadows
-			Cache.AddOrUpdate(shadowsKey, surface.Snapshot());
+			Cache.AddOrUpdate(key, surface.Snapshot());
 		}
-
-		_shadowPropertyChanged = false;
 	}
 
-	private bool ProcessContentBackgroundIfNeeded(out Color? contentBackgroundColor)
+	private static Color? GetBackgroundColor(Brush background)
```
Comment (kazo0) [inline src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs:138]:
```suggestion
	private static Color? GetBackgroundColor(Brush? background)
```

File: src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.cs [hunk 3]
```diff
@@ -29,78 +32,173 @@ public partial class ShadowContainer : ContentControl
...
+		BindToBackgroundMemberProperties(Background);
+		BindToShadowCollectionChanged(Shadows);
+
+		void OnBackgroundChanged(DependencyObject sender, DependencyProperty dp)
+		{
+			BindToBackgroundMemberProperties(Background);
+
+			InvalidateShadows();
+		}
+		void OnShadowsChanged_EX(DependencyObject sender, DependencyProperty dp)
```
Comment (kazo0) [inline src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.cs:89]: What's the _EX for?

File: src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.cs [hunk 4]
```diff
@@ -218,4 +271,10 @@ private void UpdateCanvasSize(double childWidth, double childHeight, ShadowColle
 		Canvas.SetLeft(_shadowHost, left);
 		Canvas.SetTop(_shadowHost, top);
 	}
+
+	private void InvalidateShadows(bool force = false)
+	{
+		// todo@xy: we should check for paint-states diff before proceeding
```
Comment (kazo0) [inline src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.cs:277]: Something we should address in a later PR?

**Review by kazo0 — APPROVED** [review]

**Comment by kazo0** [comment]
> it all got moved to .ShadowContainer.cs as BindToPaintingProperties and i refactored it to be cleaner and have all monitors (corner radius, background, shadows) in one place
> 
> $class-name.properties.cs should stay pure, nothing more than dependency property definitions

FYI @roubachof ^ 

Just to make sure you saw the response to your concerns

File: src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs (hunk 2 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs:138]: no need, the impl has built-in null handing logic that is needed.

**Comment by roubachof** [comment]
@kazo0 @Xiaoy312 I stand by what I said. Child CornerRadius detection has been removed and the corner radius with this PR is now directly provided by the container itself. 
It shouldn't.
The shape of the shadows should be detected from the child.
For now shadows only support rectangular shapes with corner radius, but later it will support Images, and Texts. And the shadows shape will be determined from the shape of the child, just as we doing now by detecting child corner radius.

File: src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs [hunk 5]
```diff
@@ -287,96 +182,122 @@ private static void DrawDropShadow(SKCanvas canvas, SKPaint paint, SKShadow shad
...
+			hash = hash * -1521134295 + ContentHeight.GetHashCode();
+			hash = hash * -1521134295 + PixelRatio.GetHashCode();
+			hash = hash * -1521134295 + Background.GetHashCode();
+			hash = hash * -1521134295 + CornerRadius.GetHashCode();
+			foreach (var item in ShadowInfos)
+			{
+				hash = hash * -1521134295 + item.GetHashCode();
+			}
+
+			return hash;
```
Comment (pictos) [inline src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs:291]: How about using the built-in `HashCode` struct from the [BCL](https://learn.microsoft.com/en-us/dotnet/api/system.hashcode?view=net-7.0&viewFallbackFrom=dotnet-plat-ext-3.1#moniker-applies-to)? If I'm not wrong this was added on netstandard 2.1
```suggestion
			var hashCode = new HashCode();

			hashCode.Add(ContentWidth);
			hashCode.Add(ContentHeight);
			hashCode.Add(PixelRatio);
			hashCode.Add(Background);
			hashCode.Add(CornerRadius);

			foreach (var item in ShadowInfos)
			{
				hashCode.Add(item);
			}

			return hashCode.ToHashCode();
```

**Comment by agneszitte** [comment]
@Xiaoy312 / @kazo0 / @jeromelaban 
None of these examples use proper ThemeResources so we test the theme switching colors (light <-> dark) in the toolkit sample app
image

**Comment by jeromelaban** [comment]
> None of these examples use proper ThemeResources so we can test the theme switching colors (light <-> dark) in the toolkit sample app (can be done in a separate following PR)

@agneszitte if it's not impacting the gallery, we can do it in a separate PR

**Comment by agneszitte** [comment]
@Xiaoy312 some runtime tests would also need to be added in a separated PR please

**Comment by agneszitte** [comment]
@Mergifyio backport legacy/3x

**Pull request merged**